### PR TITLE
Get chaos environments command

### PR DESCRIPTION
**Pull request by shivam-Purohit**
refers issue #164

Addition of new commands :
```
litmusctl get chaos-environment --project-id="" --environment-id=""
litmusctl list chaos-environment --project-id=""
```

### CODE REVIEW

**Comment by shivam-Purohit** [comment]
Suggest if further changes are required.

**Comment by shivam-Purohit** [comment]
@SarthakJain26 @Saranya-jena can I get a review?

File: pkg/cmd/get/environment.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+			for environmentID == "" {
+				utils.Red.Println("⛔ environment ID can't be empty!!")
+				os.Exit(1)
+			}
+		}
+
+		environmentList, err := environment.GetEnvironmentList(projectID, credentials);
+		if err != nil {
+			if strings.Contains(err.Error(), "permission_denied") {
+				utils.Red.Println("❌ The specified Project ID doesn't exist.")
```
Comment (SarthakJain26) [inline pkg/cmd/get/environment.go:66]: Please update the message to, "You don't have enough permissions to access this resource"

File: pkg/cmd/get/environment.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			}
+		}
+	},
+
+
+}
+func init() { 
+	GetCmd.AddCommand(ChaosEnvironmentCmd);
+	ChaosEnvironmentCmd.Flags().String("project-id", "", "Set the project-id to list Chaos Environment from a particular project.")
+	ChaosEnvironmentCmd.Flags().String("environment-id", "", "Set the environment-id to list Chaos Environment")
```
Comment (SarthakJain26) [inline pkg/cmd/get/environment.go:88]:
```suggestion
	ChaosEnvironmentCmd.Flags().String("environment-id", "", "Set the environment-id to get Chaos Environment")
```

File: pkg/cmd/get/environment.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+				os.Exit(1)
+			}
+		}
+	},
+
+
+}
+func init() { 
+	GetCmd.AddCommand(ChaosEnvironmentCmd);
+	ChaosEnvironmentCmd.Flags().String("project-id", "", "Set the project-id to list Chaos Environment from a particular project.")
```
Comment (SarthakJain26) [inline pkg/cmd/get/environment.go:87]:
```suggestion
	ChaosEnvironmentCmd.Flags().String("project-id", "", "Set the project-id to get Chaos Environment from a particular project.")
```

File: pkg/cmd/list/environment.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+			for projectID == "" {
+				utils.Red.Println("⛔ Project ID can't be empty!!")
+				os.Exit(1)
+			}
+		}
+
+		environmentList, err := environment.GetEnvironmentList(projectID, credentials);
+		if err != nil {
+			if strings.Contains(err.Error(), "permission_denied") {
+				utils.Red.Println("❌ The specified Project ID doesn't exist.")
```
Comment (SarthakJain26) [inline pkg/cmd/list/environment.go:54]: Please update the message to, "You don't have enough permissions to access this resource"

File: pkg/cmd/list/list.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"github.com/spf13/cobra"
+)
+
+// getCmd represents the get command
+var ListCmd = &cobra.Command{
+	Use: "list",
+	Short: `Examples:
+		#get list of chaos Chaos Experiments
```
Comment (SarthakJain26) [inline pkg/cmd/list/list.go:26]:
```suggestion
		#get list of chaos Chaos Environments
```

**Review by SarthakJain26 — CHANGES_REQUESTED** [review]

**Comment by SarthakJain26** [comment]
@shivam-Purohit can you please also share the screen shots of the command and the output

**Comment by SarthakJain26** [comment]
@shivam-Purohit please add these commands in the readme and usage docs as well

**Comment by shivam-Purohit** [comment]
> @shivam-Purohit can you please also share the screen shots of the command and the output

sure!
ubuntuMachine  Running  - Oracle VM VirtualBox 29-12-2023 10_17_17
ubuntuMachine  Running  - Oracle VM VirtualBox 29-12-2023 10_16_26

**Comment by shivam-Purohit** [comment]
> @shivam-Purohit please add these commands in the readme and usage docs as well

I added these in 0.23.0. Do we need to add in any other versions? i think we do not need that.

@SarthakJain26 I fixed all the changes you requested. Let me know if further changes are required.

**Comment by SarthakJain26** [comment]
@shivam-Purohit , you have displayed the complete response coming from the API which is not a very good user experience. You will have to extract required fields and you can display them in table format. I have attached the screenshot for reference. You can display environmentID, name, createdAt and createdBy fields. Please update the PR accordingly
Screenshot 2023-12-29 at 11 24 06 AM

**Comment by shivam-Purohit** [comment]
hey @SarthakJain26 I am currently getting the response as null in many fields which ig should not. I tried rechecking the API. Is there something I am doing here to retrieve the data?

**Comment by SarthakJain26** [comment]
> hey @SarthakJain26 I am currently getting the response as null in many fields which ig should not. I tried rechecking the API. Is there something I am doing here to retrieve the data?

@shivam-Purohit Possibly yes. You need to check on what is the graphql request you are sending and then what is the data you are expecting back. While decoding the response from the graphql server you need to check if the struct you are using to decode the response has the required fields or not.

**Comment by shivam-Purohit** [comment]
@SarthakJain26 the error was actually due to query itself which we were sending through api. I changed it a bit for the newcommand. 
I also refactored the output and reused some other components for similarity and reusability.
ubuntuMachine  Running  - Oracle VM VirtualBox 07-01-2024 18_21_19

**Comment by SarthakJain26** [comment]
Thanks @shivam-Purohit, update output looks awesome

**Comment by Nageshbansal** [comment]
@SarthakJain26, do we really need two separate commands for this, i.e., list and get? I think  `get` should be enough only to show a list of all environments, as that is what we're doing for experiments and infra.

**Comment by SarthakJain26** [comment]
> @SarthakJain26, do we need two separate commands for this, i.e., list and get? Maybe `get` should be enough only to show a list of all environments, as that is what we're doing for experiments and infra.

@Nageshbansal , by convention, list command can be used to list the resources under the project with required info, and get command can be used to get additional info about a particular resource. We can follow the same in other places as well. 
Adding to this @shivam-Purohit please let us know if you can add other details also in the get environment command in couple of days so that we can merge this PR and take it in this release.

**Comment by shivam-Purohit** [comment]
> > @SarthakJain26, do we need two separate commands for this, i.e., list and get? Maybe `get` should be enough only to show a list of all environments, as that is what we're doing for experiments and infra.
> 
> @Nageshbansal , by convention, list command can be used to list the resources under the project with required info, and get command can be used to get additional info about a particular resource. We can follow the same in other places as well. Adding to this @shivam-Purohit please let us know if you can add other details also in the get environment command in couple of days so that we can merge this PR and take it in this release.


```
type Environment struct {
	ProjectID     string          `json:"projectID"`
	EnvironmentID string          `json:"environmentID"`
	Name          string          `json:"name"`
	Description   *string         `json:"description"`
	Tags          []string        `json:"tags"`
	Type          EnvironmentType `json:"type"`
	CreatedAt     string          `json:"createdAt"`
	CreatedBy     *UserDetails    `json:"createdBy"`
	UpdatedBy     *UserDetails    `json:"updatedBy"`
	UpdatedAt     string          `json:"updatedAt"`
	IsRemoved     *bool           `json:"isRemoved"`
	InfraIDs      []string        `json:"infraIDs"`
}
```

these are the details we can get from the response. @SarthakJain26 let me know what details you want to add, imo we can add type, description or maybe updated at and updated by.

**Comment by shivam-Purohit** [comment]
> @shivam-Purohit you can also add the InfraIDs along with what you mentioned so that it infras within the environment can be linked.

sure!

**Review by SarthakJain26 — APPROVED** [review]

**Comment by SarthakJain26** [comment]
@shivam-Purohit can you please fix the fmt issues due to which build pipeline is failing and also sign your commits so that DCO can pass. We can then merge the PR

**Comment by shivam-Purohit** [comment]
> can you please fix the fmt issues due to which build pipeline is failing and also sign your commits so that DCO can pass. We can then merge the PR

should I squash the commits into one?

**Comment by Nageshbansal** [comment]
@shivam-Purohit Squashing the commit is not required ( it will be taken care of while merging) , for now please fix the fmt in go files ( you can use gofmt command to check which files are not correctly formatted)

**Comment by Nageshbansal** [comment]
@shivam-Purohit Squashing the commit is not required ( it will be taken care of while merging) , for now please fix the fmt in go files ( you can use gofmt command to check which files are not correctly formatted)

**Comment by Nageshbansal** [comment]
For DCO, you can refer to : https://github.com/litmuschaos/litmus/blob/master/CONTRIBUTING.md#sign-your-work-with-developer-certificate-of-origin

**Comment by shivam-Purohit** [comment]
@Nageshbansal my first commit was unsigned i could not find any other way to sign it. So I merged it. If the workflows get approval i can test the changes i just made . I use gofmt -w . that should be able to get rid of the fmt error

**Comment by shivam-Purohit** [comment]
> @shivam-Purohit can you please fix the fmt issues due to which build pipeline is failing and also sign your commits so that DCO can pass. We can then merge the PR

@SarthakJain26 can you run the workflow again ig this should be fine now at least the fmt error

File: pkg/cmd/get/environment.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+			}
+		}
+		environmentListData := environmentList.Data.ListEnvironmentDetails.Environments
+		writer := tabwriter.NewWriter(os.Stdout, 30, 8, 0, '\t', tabwriter.AlignRight)
+		writer.Flush()
+		for i := 0; i < len(environmentListData); i++ {
+			if environmentListData[i].EnvironmentID == environmentID {
+				intUpdateTime, err := strconv.ParseInt(environmentListData[i].UpdatedAt, 10, 64)
+				if err != nil {
+					fmt.Println("Error converting UpdatedAt to int64:", err)
```
Comment (Jonsy13) [inline pkg/cmd/get/environment.go:82]: utils.Red.Println

File: pkg/cmd/get/environment.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		utils.PrintError(err)
+
+		projectID, err := cmd.Flags().GetString("project-id")
+		utils.PrintError(err)
+
+		if projectID == "" {
+			utils.White_B.Print("\nEnter the Project ID: ")
+			fmt.Scanln(&projectID)
+
+			for projectID == "" {
```
Comment (Jonsy13) [inline pkg/cmd/get/environment.go:46]: Can we use `if` here?

File: pkg/cmd/get/environment.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+		}
+
+		environmentID, err := cmd.Flags().GetString("environment-id")
+		utils.PrintError(err)
+
+		if environmentID == "" {
+			utils.White_B.Print("\nEnter the Environment ID: ")
+			fmt.Scanln(&environmentID)
+
+			for environmentID == "" {
```
Comment (Jonsy13) [inline pkg/cmd/get/environment.go:59]: Can we use if here?

File: pkg/cmd/get/environment.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+			if environmentListData[i].EnvironmentID == environmentID {
+				intUpdateTime, err := strconv.ParseInt(environmentListData[i].UpdatedAt, 10, 64)
+				if err != nil {
+					fmt.Println("Error converting UpdatedAt to int64:", err)
+					continue
+				}
+				updatedTime := time.Unix(intUpdateTime, 0).String()
+				intCreatedTime, err := strconv.ParseInt(environmentListData[i].CreatedAt, 10, 64)
+				if err != nil {
+					fmt.Println("Error converting CreatedAt to int64:", err)
```
Comment (Jonsy13) [inline pkg/cmd/get/environment.go:88]: utils.Red.Println

**Review by Saranya-jena — APPROVED** [review]

File: pkg/cmd/get/environment.go (hunk 7 above)
Comment (shivam-Purohit) [reply pkg/cmd/get/environment.go]: I picked it from experiments get command so the same code is repeated there but doesn't make any sense to put for and then exit anyway. If would be a better choice

**Comment by shivam-Purohit** [comment]
@Jonsy13 @Saranya-jena I made the changes suggested you can take a look. Suggest if any further changes need to be done.

**Comment by SarthakJain26** [comment]
@shivam-Purohit please fix the build pipeline

**Comment by shivam-Purohit** [comment]
@SarthakJain26 fixed the fmt error. Hope it resolves the build workflow

**Comment by shivam-Purohit** [comment]
Looks like it worked!

**Pull request merged**